src/irmin-pack/unix/traverse_pack_file.ml

```diff
@@ -211,29 +226,77 @@ end = struct
     | Invalid_argument msg when msg = "String.blit / Bytes.blit_string" ->
         raise Not_enough_buffer
 
+  (* duplicated code with irmin-tezos-utils *)
+  let decode_entry_len buffer =
+    let buffer = Bytes.unsafe_to_string buffer in
+    let i0 = 0 in
+    let ilength = i0 + Hash.hash_size + 1 in
+    let pos_ref = ref ilength in
+    let suffix_length = Varint.decode_bin buffer pos_ref in
+    let length_length = !pos_ref - ilength in
+    Hash.hash_size + 1 + length_length + suffix_length
+
   (* Read at most [len], by checking that [(off, len)] don't go out of bounds of
      the suffix file. *)
-  let io_read_at_most ~off ~len b suffix =
-    let bytes_after_off =
-      let open Int63.Syntax in
-      File_manager.Suffix.end_offset suffix - off
-    in
+  let io_read_at_most ~off ~len buffer dispatcher =
     let len =
       let open Int63.Syntax in
+      let bytes_after_off = Dispatcher.end_offset dispatcher - off in
       if bytes_after_off < Int63.of_int len then Int63.to_int bytes_after_off
       else len
     in
-    File_manager.Suffix.read_exn suffix ~off ~len b;
-    len
+    let accessor = Dispatcher.create_accessor_exn dispatcher ~off ~len in
+    Dispatcher.read_exn dispatcher accessor buffer
+
+  let on_entry { data; key } stats iter_pack_entry missing_hash =
+    [%log.debug "k = %a (off, len, kind) = %a" pp_key key pp_index_value data];
+    match iter_pack_entry key data with
+    | Ok () -> Option.map Fun.id missing_hash
+    | Error `Inconsistent_entry ->
+        Stats.duplicate_entry stats;
+        Option.map Fun.id missing_hash
+    | Error (`Missing_hash x) ->
+        Stats.missing_hash stats;
+        Some x
 
-  let ingest_data_file ~initial_buffer_size ~progress ~total suffix
+  let ingest_data_file_after_v3 ~initial_buffer_size ~progress dispatcher
       iter_pack_entry =
+    let stats = Stats.empty () in
+    let max_bytes_needed_to_discover_length =
+      Hash.hash_size + 1 + Varint.max_encoded_size
+    in
+    let min_bytes_needed_to_discover_length = Hash.hash_size + 1 in
```

Comment:
nit: you could reorder these lines to make the relationship between min and max explicit. eg something like

```suggestion
    let min_bytes_needed_to_discover_length = Hash.hash_size + 1 in
    let max_bytes_needed_to_discover_length =
      min_bytes_needed_to_discover_length + Varint.max_encoded_size
    in
```